feat: support via parameter

marius-at-atb opened the pull request:
Support via parameter
Resolves https://github.com/AtB-AS/kundevendt/issues/2732
Related app PR: https://github.com/AtB-AS/mittatb-app/pull/3926

Acceptance criteria
- [ ] The via parameter should be included by all non-deprecated API endpoints
- [ ] Favourites should work for both new and old app versions -> whether used with the old lineName or new via parameter
- [ ] Tests should pass


Special thanks to @rosvik for good help!

rosvik commented on src/service/impl/departures-grouped/utils/grouping.ts:
```diff
@@ -134,7 +146,9 @@ export default function mapQueryToGroups(
           const lineInfo: DepartureLineInfo = {
-            lineName: lineInfoEntry.destinationDisplay?.frontText ?? '',
+            lineName: mapToLegacyLineName(lineInfoEntry.destinationDisplay),
+            destinationDisplay:
+              ensureViaFormat(lineInfoEntry.destinationDisplay) ?? {},
             lineNumber: lineInfoEntry.serviceJourney?.line.publicCode ?? '',
```
Ref. discussion, maybe:

```ts
const lineInfo: DepartureLineInfo = {
  lineName: mapToLegacyLineName(lineInfoEntry.destinationDisplay),
  destinationDisplay: lineInfoEntry.destinationDisplay ?? {},
  lineNumber: lineInfoEntry.serviceJourney?.line.publicCode ?? '',
  transportMode: lineInfoEntry.serviceJourney?.line.transportMode,
  transportSubmode:
    lineInfoEntry.serviceJourney?.line.transportSubmode,
  quayId: quay.id,
  notices: (lineInfoEntry.notices as Notice[]) ?? [],
  lineId: lineInfoEntry.serviceJourney?.line.id ?? '',
};

if (!isFavorite(ensureViaFormat(lineInfoEntry.destinationDisplay) ?? {},, stopPlaceInfo.id)) {
  continue;
}
```

gorandalum commented:
Ref discussion. The plan now is:
- Not map/modify data in the api from app to BFF. That means no splitting-logic on the via parameter.
- The migration in stored favorites puts the old linename just into front text in the new format. No splitting-logic on the via parameter.
- All splitting-logic or concatenate-logic regarding front text and via is done in the comparison function in the bff.

marius-at-atb commented:
Both this PR and https://github.com/AtB-AS/mittatb-app/pull/3926 are now updated and implemented according to [plan](https://github.com/AtB-AS/atb-bff/pull/301#issuecomment-1763890648)

gorandalum reviewed (approved):
Nice!

gorandalum commented on src/service/impl/departures/utils/__tests__/destination-display-matching.test.ts:
I really like these tests! Here are some thoughts on what also can be done:

- I think inlining the destination display values in each test is better than having variables for them. Even though it gives some more duplication, I think it's easier/faster to read the tests that way.
- The `frontText` and `via` can be `undefined` according to the Entur doc. Maybe add tests for that?
- In addition to testing splitting on "-", could we add some tests for " - ", "/" and " / "?

marius-at-atb replied:
Yes, we can definitely add tests for those cases as well and inline the destinationDisplays :+1:

rosvik commented on src/service/impl/departures/utils/favorites.ts:
```diff
@@ -79,3 +82,47 @@ export function filterStopPlaceFavorites(
+  );
+}
+
+/* NB this is the same function as in the app. Keep in sync! */
```
Will this be an issue if the app changes something in a future version, and it becomes out of sync with the BFF?

marius-at-atb replied:
"Keep in sync" was maybe not the best comment here.
`destinationDisplaysAreEqual` is just a simple util for checking if `destinationDisplay1 === destinationDisplay2`, without any "smart" behaviour.
So it would not be of great concern

marius-at-atb replied:
What's important is what we send in to this function

rosvik reviewed (approved):
Really good work 💯 Had a couple of more questions, but looks good overall!

marius-at-atb commented on src/service/impl/departures-grouped/utils/grouping.ts:
```diff
@@ -134,7 +143,8 @@ export default function mapQueryToGroups(
 
           const lineInfo: DepartureLineInfo = {
-            lineName: lineInfoEntry.destinationDisplay?.frontText ?? '',
+            lineName: mapToLegacyLineName(lineInfoEntry.destinationDisplay),
```
Your understanding is correct!
I'll add a comment :+1:

marius-at-atb commented on src/service/impl/departures/utils/converters.ts:
```diff
@@ -0,0 +1,32 @@
+    return {frontText, via: undefined};
+  }
+
+  const via = viaItemsString.split(/\s*[/,-]\s*/); // split on "/" and "-" and trim
```
Atm this is only a best estimate, not something we know for sure yet. Might have to come back to this function (and one of the test cases) later..
